transport-native-unix-common/src/main/java/io/netty/channel/unix/UnixChannelUtil.java

```diff
@@ -29,6 +29,6 @@ private UnixChannelUtil() {
      * (We check this because otherwise we need to make it a new direct buffer.)
      */
     public static boolean isBufferCopyNeededForWrite(ByteBuf byteBuf) {
-        return !(byteBuf.hasMemoryAddress() || byteBuf.isDirect() && byteBuf.nioBufferCount() <= IOV_MAX);
+        return !byteBuf.hasMemoryAddress() || !(byteBuf.isDirect() && byteBuf.nioBufferCount() <= IOV_MAX);
```

Comment:
IMHO the following is a bit simpler: `!byteBuf.hasMemoryAddress() || !byteBuf.isDirect() || byteBuf.nioBufferCount() > IOV_MAX` .. each term can be evaluated independently and no need to negate a group.